### ISSUE TITLE
[DYNAREC] Introducing differential testing to box64 JIT

### ISSUE THREAD

**Issue by xctan** [ptitSeb/box64#634]
When porting dynarec to RISC-V, @ksco and I found that many long and less frequent x86_64 opcodes are not well tested, so we are unaware of hidden bugs inside. By introducing fuzzing tests like csmith, we have more chance to detect such bugs. But the problem with csmith is that it generates very few 8-bit and 16-bit math operations, so related opcodes are rarely tested. I suggested to use sed to replace `int32_t` with `int8_t`, making gcc produce a bunch of unimplemented opcodes, but this breaks the checksum mechanism of csmith. In order to verify the correctness of JIT implementation, we probably need to compare x86_64 registers between dynarec and the interpreter, side by side, for each translated instruction. With differential testing infrastructure, we can better debug dynarec and find bugs easily and precisely.

**Comment by ksco** [comment]
If it can be made self-checking, it can also help users discover and report issues. For example, allowing users to start the diff test mode from a certain PC, in this mode, box64 becomes purely interpreted, but at the same time, single-instruction JIT is performed. After the execution is completed, there is no write-back, but a comparison is made with the execution result of the interpreter.

**Comment by ptitSeb** [comment]
So, I have made something.
It's pretty complex but seems to work fine.
Basicaly, each instruction in the dynarec is compared to what the interpreter will do. this is done by running the interpretor befaore the dynarec opcode, on a secondary x64emu_t structure, but redirecting all memory write to a buffer.
After the dynarec opcode execute, all regs and the eventual memory is compared and difference are printed out.
This is the simplified version, as there are many special case of course.

I tested it on ARM64 and RV64 and seems to work fine. It might still have a few ugs here and there, it has been quickly tested.

To activate it, use `BOX64_DYNAREC_TEST=1` and it will print out difference if there are.
Try witht the `test10` as it will most likely print difference in memory. This is due the threads running on this one, the differences are because another thread change the memory between the interpreter step and the dynarec one. Not much can be done here....
